Removed use of deprecated std::aligned_storage

Dr15Jones opened the pull request:
#### PR description:

The type was deprecated in C++23.

#### PR validation:

Compiling in the C++23 work area no longer shows the warning. Running valgrind on the unit test for the SoATuple did not show any problems.

resolves https://github.com/cms-sw/framework-team/issues/1761

Dr15Jones committed:
Removed use of deprecated std::aligned_storage

The type was deprecated in C++23. Switch to using new with alignment specifier.

Dr15Jones committed:
Use std::byte for storage type in edm::SoATuple

Dr15Jones committed:
Use std::assume_aligned instead of compiler specific function

Dr15Jones committed:
Must call explicit alignment form of delete

Dr15Jones committed:
Modernized SoATupleHelper

- switched from recursive types to fold expressions with lambdas
- use constexpr
- switch to using expressions

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-49712/47301

Dr15Jones commented:
Please test

Dr15Jones committed:
switched to explicit operator new

- being symmetric with the delete case was viewed as desireable based on PR discussion
- allocate only what is needed without padding at end of array as we are now allocating just bytes rather than a type using max_alignment size.

cmsbuild commented:
+1

**Size**: This PR adds an extra 24KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d9bef6/50455/summary.html
**COMMIT**: 4eb084f32d33e25c6379c86c7ee2ccdf627b224d
**CMSSW**: CMSSW_16_1_X_2026-01-06-1100/el8_amd64_gcc13
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/49712/50455/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - ROOTFileChecks: Some differences in event products or their sizes found
 - [Reco comparison results: 0 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_16_1_X_2026-01-06-1100+d9bef6/72591/validateJR.html)
 - Reco comparison had  4 failed jobs
 - DQMHistoTests: Total files compared: 53
 - DQMHistoTests: Total histograms compared: 4280553
 - DQMHistoTests: [Total failures: 13](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_16_1_X_2026-01-06-1100+d9bef6/72591/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4280520
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 52 files compared)
 - Checked 227 log files, 198 edm output root files, 53 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_16_1_X_2026-01-06-1100+d9bef6/72591/triggerResults): no differences found

makortel commented:
+core

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @sextonkennedy, @mandrenguyen, @ftenchini (and backports should be raised in the release meeting by the corresponding L2)

mandrenguyen commented:
+1
